MaterialLib/MPL/Properties/CapillaryPressureSaturation/CapillaryPressureVanGenuchten.cpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * \file
+ * \copyright
+ * Copyright (c) 2012-2020, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ * Created on March 20, 2020, 9:59 AM
+ */
+
+#include "CapillaryPressureVanGenuchten.h"
+
+#include <algorithm>
+#include <cmath>
+
+#include "MaterialLib/MPL/Medium.h"
+
+namespace MaterialPropertyLib
+{
+CapillaryPressureVanGenuchten::CapillaryPressureVanGenuchten(
+    double const residual_liquid_saturation,
+    double const residual_gas_saturation,
+    double const exponent,
+    double const p_b,
+    double const maximum_capillary_pressure)
+    : _S_L_res(residual_liquid_saturation),
+      _S_L_max(1. - residual_gas_saturation),
+      _m(exponent),
+      _p_b(p_b),
+      _p_cap_max(maximum_capillary_pressure)
+{
+    if (!(_m > 0 && _m < 1))
+    {
+        OGS_FATAL(
+            "The exponent value m = {:g} of van Genuchten capillary pressure "
+            "model, is out of its range of (0, 1)",
+            _m);
+    }
+}
+
+PropertyDataType CapillaryPressureVanGenuchten::value(
+    VariableArray const& variable_array,
+    ParameterLib::SpatialPosition const& /*pos*/, double const /*t*/,
+    double const /*dt*/) const
+{
+    double const S_L = std::get<double>(
+        variable_array[static_cast<int>(Variable::liquid_saturation)]);
+
+    if (S_L <= _S_L_res)
+    {
+        return _p_cap_max;
+    }
+
+    if (S_L >= _S_L_max)
+    {
+        return 0;
+    }
+
+    double const S_eff = (S_L - _S_L_res) / (_S_L_max - _S_L_res);
+    assert(0 <= S_eff && S_eff <= 1);
+
+    double const p_cap =
+        _p_b * std::pow(std::pow(S_eff, -1.0 / _m) - 1.0, 1.0 - _m);
+    assert(p_cap > 0);
+    return std::min(p_cap, _p_cap_max);
+}
+
+PropertyDataType CapillaryPressureVanGenuchten::dValue(
+    VariableArray const& variable_array, Variable const primary_variable,
+    ParameterLib::SpatialPosition const& /*pos*/, double const /*t*/,
+    double const /*dt*/) const
+{
+    (void)primary_variable;
+    assert((primary_variable == Variable::liquid_saturation) &&
+           "CapillaryPressureVanGenuchten::dValue is implemented for "
+           "derivatives with respect to liquid saturation only.");
+
+    double const S_L = std::get<double>(
+        variable_array[static_cast<int>(Variable::liquid_saturation)]);
+
+    if (S_L <= _S_L_res)
```

Comment:
```suggestion
    if (!(S_L > _S_L_res))
```
in order to avoid the identical comparison of two real numbers.

Reply:
Interesting. Do you have an example where
```c++
assert((a <= b) == (!(a > b)));
```
fails? I [tried](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:c%2B%2B,selection:(endColumn:3,endLineNumber:26,positionColumn:3,positionLineNumber:26,selectionStartColumn:3,selectionStartLineNumber:26,startColumn:3,startLineNumber:26),source:'%23include+%3Climits%3E%0A%23include+%3Ciostream%3E%0A%0Adouble+g(double+a,+double+b)%0A%7B%0A++++if+(a+%3E%3D+b)%0A++++return+1%3B%0A++++else%0A++++return+0%3B%0A%7D%0A%0Adouble+f(double+a,+double+b)%0A%7B%0A++++if+(!!(a+%3C+b))%0A++++return+1%3B%0A++++else%0A++++return+0%3B%0A%7D%0A/*%0Aint+main()%0A%7B%0A++++double+a,+b%3B%0A++++std::cout+%3C%3C+0.*g(a,+b)+%3C%3C+%22+%22+%3C%3C+0.+*+f(a,+b)+%3C%3C+%22%5Cn%22%3B%0A++++return+0%3B%0A%7D%0A*/'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:30.773018217626785,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:vcpp_v19_24_x64,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:c%2B%2B,libs:!(),options:/Ox,selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'x64+msvc+v19.24+(Editor+%231,+Compiler+%231)+C%2B%2B',t:'0')),k:35.89364844903989,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,wrap:'1'),l:'5',n:'0',o:'%231+with+x64+msvc+v19.24',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) to find difference with different compilers and optimization levels and there are some.
For gcc-9.3.0 and msvc 19.24 and optimization >= O1 the comparisons are using the same instruction `comisd`: https://www.felixcloutier.com/x86/comisd
clang-10.0 and optimization >= 01 yields "different" instructions `cmpltsd` `cmpnlesd`, which, to my understanding, are just so called pseudo-op for `cmpsd`
```
CMPLTSD xmm1, xmm2	CMPSD xmm1, xmm2, 1
```
where the last 1 indicates less-then predicate. https://www.felixcloutier.com/x86/cmpsd

`comisd` sets multiple flags depending on the comparison. `cmpsd` set's destination register to 0xFF or 0x00 depending on the outcome.

As I do understand how the floating point numbers are compared the above assertion can never fail. Please show a case, where it fails.

Reply:
Nice explanation. Go forward.

Reply:
- IMHO the version `x <= y` is much more readable than `!(x > y)`. Therefore I would keep the code as it is. 
- As far as I remember Scott Meyers recommended to implement the operator `<` using the negation of the operator `<=`. So the above expression should be identical.

Reply:
@wenqing Please show a case where this relation operators do fail for doubles.

Reply:
For example, `S_L=NaN`. You can also use `std::islessequal` and `std::isgreaterequal`. For float number comparison, IMHO, operator of `==`, `<=`, `>=` should be avoided.

Reply:
If you want to keep `if (S_L <= _S_L_res)`,  please go forward. `S_L=NaN` seldom occurs, even   in the equation with saturation as the primary variable. Anyway, the program can be halt by linear/non-linear solvers under such case.

Reply:
I see how that example is showing inconsistency. The reason for NaN comparisons to be different is that **any** comparison with NaN is `false` but the "not-equal". Then negating the result is not really making it correct, though the expression becomes `true`. The `std::isgreaterequal(a,b)` is also not equal to `!(a < b)` for NaN's, beacause `(a < b)` for a NaN is always `false`.
If this would restore some logic, yes, I'd agree that the expression with negation is superior.
C'mon, NaN is representation of an *undefined value*, how does the negated version save the logic?
To catch runtime errors, use `ogs --enable-fpe x.prj`.

From `NaN < b` being `false` you cannot conclude that `b > NaN` is `true`, the latter is also `false`.